### ISSUE TITLE
Changing the default language will cause existing translations to break

### ISSUE THREAD

**Issue by iSaadSalman** [rainlab/translate-plugin#231]
I have setuped by custom plugin using translate plugin and information was entered for both languages. But once I switch the default language from English to Arabic the arabic word are in english and there are no words for english

**Comment by Minius-dev** [comment]
Yes, it`s bug i think. In my websites to. I can`t change default language later. 
I can say, that octobercms it`s very nice system, but there is a problem with translations in some cases... and it would be nice to get some updates from the team

**Comment by pbsmultimedia** [comment]
+1

**Comment by LukeTowers** [comment]
Closing as it has been over a month since any activity on this occurred.

**Comment by LukeTowers** [comment]
@munxar @vannut Are you guys able to see what causes this issue of translations breaking when the default language for a site is changed?

**Comment by munxar** [comment]
@LukeTowers tl;dr it works as implemented.
I'll try to explain why:
If you use the ModelTranslation behavior it is applied to a model that is related to a database table.
The values stored in this table are assumed to be in default language.
If you add a translatable column via $translatable = [] and only have the default language, basically nothing happens in the rainlab_translate_attributes or rainlab_tranlsate_indexes (translation tables) tables (because the data is in the table itself).
If you add another language as non default and write a translatable value to this model, it will store the data in the translation tables.
If you switch now default language, the data still in the models database tables is now default.
Because there where no entries for the former default language the translations are now missing. there still in the db because if you switch everything back, it works like expected.

long story short, here are two solution that come to my mind: 

a) choose your default language before entring data

b) if you have to switch, write a migration script that flips model <-> translated data for this to languages (could be a optional feature in the translate plugin)

hope this is clear, so no real bug, just the normal behavior with model translation.

**Comment by joelamltc** [comment]
Would it be good if store all translation in the rainlab_translate_attributes table?
Then no matter how you switch it will still can grab the right translations.

**Comment by munxar** [comment]
@joelamltc it depends on your definition of 'good' :-)
well, I think no, because the plugin is build in a way that you can just add or remove it as pleased.
it doesn't alter data in this process, which is a good thing.
Imho we should instead add a feature to copy translation data from one locale to another.
that would be:
a) optional / non destructive
b) solve the original problem (with a little hand work, but I think its fair because switching the default locale is not something you usually do in a productive environment, do you?!)

**Comment by joelamltc** [comment]
yea I totally understand.

for b) in production server no, but sometime in development will do. Coz sometimes the client didn't know what language they should default, and I just always default to en, but somehow they want to change it to others. Then, it would be my problems if can't switch or maybe just like you said, do some little handy work to migrate it.

**Comment by LukeTowers** [comment]
@munxar what do you think about triggering the copying of default lang attributes (i.e. the language swap) when changing the default locale?

**Comment by munxar** [comment]
@LukeTowers wouldn't do that. Why?
1. We don' t know the side effects of a model save, or what hocks are triggered.
In the worst case data is lost. You wouldn't expect a data modification by switching the default language, would you?

2. What will happen, if the new default language doesn't have a value, but the db requires one?
The translated fields don't have have any validation rules or schema. An exchange is not always possible.

**Comment by LukeTowers** [comment]
Fair enough. How about as a console command instead then?

**Comment by munxar** [comment]
@LukeTowers I'll try to be more clear:
It's very hard to do, because we need to
1. know all translated model classes
2. iterate over their data and write it as normal translations into the translation tables
3. get the data from the translation tables and write them into the tables (as default)

I have no clue how to solve point 1. point 2. is relatively easy and 3. is technically impossible.

Let's make an example, default language is *en* and a second language *de* (german), and we have a table **foo_bar_items** schema like this:

| id | name |
| --- | --- |
| 1 | hello |
| 2 | world |

our **rainlab_translate_attributes** table is empty:

| id | locale | model_id | model_type | attribute_data |
| --- | --- | --- | --- | --- |

let's translate the name of the record with id 1 to *de* (german):

| id | locale | model_id | model_type | attribute_data |
| --- | --- | --- | --- | --- |
| 1 | de | 1 | Foo\Bar\Models\Item | { "name": "hallo" } |

Now let's decide that our default language is german *de* instead of english *en*:
But how should we copy the data? Let's be naive and iterate over the rainlab_translate_attributes table and swap the data:

| id | name |
| --- | --- |
| 1 | hallo |
| 2 | world |

| id | locale | model_id | model_type | attribute_data |
| --- | --- | --- | --- | --- |
| 1 | en | 1 | Foo\Bar\Models\Item | { "name": "hello" } |

The translate table contains correct data, but our items table has now languages mixed up which doesn't make sense.
Another question is: What if there is a model without any translations made yet? how do we know about it and fill the rainlab_translate_attributes table?

**Comment by LukeTowers** [comment]
@munxar you're right, this is technically very difficult to pull off. Could you recommend an addition to the documentation that would warn developers to not go around changing the default language for these reasons?

**Comment by mjauvin** [comment]
Maybe we could add a confirmation popup when changing the default locale, warning people about this?

**Comment by CptMeatball** [comment]
So basically there is no solution for this problem (yet)? 

@LukeTowers: Warning developers not changing the default language is one step. But developers aren't always in the possibility to make this choice. If the client wants to change the base language and the website is relatively large, nobody's gonna be happy 🤔 

A dirty fix would be the make it a part of the front-end component, but that would "break" a site when removing the plugin. It would change languages unexpectedly

**Comment by joelamltc** [comment]
Or maybe we can actually consider make translate plugin as part of OctoberCMS, make it translate ready. When user / developer want it, it already there, otherwise, just let it rest inside OctoberCMS when translation is not needed.

I am not recommending sth, just pure discussion.

**Comment by bennothommo** [comment]
@joelamltc The Translate plugin is a RainLab plugin, which means it is maintained by the core developers and should be treated as "core" functionality, but it won't be merged into the October CMS core as it really is supposed to be an optional augmentation to an October CMS site. A lot of people only work with one language, and have no need for translations.

**Comment by mjauvin** [comment]
@daftspunk did you figure out a way to address this?

**Comment by zakir1929** [comment]
public function boot(){
        if(!$this->app->runningInBackend()) {
            $translator = \RainLab\Translate\Classes\Translator::instance();
            if(!\Session::get(\RainLab\Translate\Classes\Translator::SESSION_LOCALE)){
                $translator->setLocale('en',TRUE);
            }
        }
    }

I have to set the default language for front-end only so I have written that code and it works perfectly !!!

**Comment by foxik0169** [comment]
Hello, 

I think the problem can be solved just by adding one more setting which will be the 'default' frontend language. Similarly to what @zakir1929 did, just as a setting. In the backend, the translations would still work as now (one language is default, and translations are stored in separate table). 

Also I think this is a naming problem, cause the 'default' now means two things:
 1. Default in the frontend - actually Default language?
 2. Language of the non-translated stored content - something more like Source language?

I understand that this might not be totally user-friendly but I guess for developers it makes sense?..

Daniel

**Comment by TomeuMut** [comment]
Hi, The answer switch locale in BD isn't work for me, someone get different answer for this bug?

**Comment by mjauvin** [comment]
@TomeuMut did you try what was suggested here:
https://github.com/rainlab/translate-plugin/issues/231#issuecomment-658899574

**Comment by TomeuMut** [comment]
@mjauvin Yes, but not working... One Idea I think is in folder plugin create a file routes.php and save the values the model in BD and then change de lang...

**Comment by TomeuMut** [comment]
Hi, I found one solution for this problem, I create two functions.

```
function saveLang($page){
        /*
        * For Steps, Step 1
        * Get lang Values and get aux lang
        * And Set default lang to aux
        */

        $page->translateContext('ca');
        $attributesCa = $page->toArray();

        $page->translateContext('es');
        $attributesEs = $page->toArray();

        $page->translateContext('en');
        updateModel($page,$attributesCa);
        $attributesEn = $page->toArray();
    }

    function changeLang($page){
        /*
         * Step 3, Set Values in fields lang.
         */
        $page->translateContext('en');
        $attributesEn = $page->toArray();

        $page->translateContext('ca');
        $attributesCa = $page->toArray();

        $page->update($attributesEn);

        $attributesCaUpdated = $page->toArray();

        $page->translateContext('es');
        $attributesEs = $page->toArray();

        $page->update($attributesCa);
        $attributesEs = $page->toArray();
    }
```
The Step 2 is change default lang manual because make default lang not work with code... and I use the EN for lang aux. 

This is a litlle Patch but is something.

**Comment by tombien** [comment]
We found problem in Static Pages. We have EN and FR languages for example. FR is default. Static Pages makes `themes/my-theme/content/static-pages` and `themes/my-theme/content/static-pages-en`. If you switch default language from FR to EN, site is going to break, because folders did not renamed so you have folder with french for default (EN) language and for french you have nothing.

**Comment by bennothommo** [comment]
@tombien Would you mind creating a separate ticket for this issue, as I believe yours might be a Pages plugin issue instead?

**Comment by ivang76** [comment]
Hi guys,
I'm still facing with same issue (the last one written by @tombien )... do you know if are there updates or if exists already a Pages ticket?

**Comment by daftspunk** [comment]
Unfortunately, this is a known issue with the plugin: you must select the default language before performing any translations. We hope to have this resolved in the next iteration of October's translation capabilities

**Comment by ivang76** [comment]
Hi @daftspunk I'd need your help if possible.

If in a project the default language changes, the old url set at the begin of the cms pages, would need to be changed as well, but in some case the user could not change manually that value (in my project this is due to hidden operations depending by editor role for example).

So I need that the system get the localeUrl[?] everytime and dosn't trust to the default value.
To do this I changed the function _rewriteTranslatablePageUrl_ in the file _TranslatablePageUrl.php_ because I need to execute always this command for all the CMS pages:
**$localeUrl = $this->getSettingsUrlAttributeTranslated($locale) ?: $localeUrl;**

This just works but I don't like ofcourse to work on the plugin code itself and I would "force" this behaviour someway in my custom Plugin.php. Is this possible?

thank you!

**Comment by daftspunk** [comment]
A forceDefault method has been added for protection against this. See ec7121b158f465ccbd07f0f427ca6a06039c971c

```
RainLab\Translate\Models\Locale::forceDefault('en');
```

This should help. Let me know otherwise

**Comment by ivang76** [comment]
thank you @daftspunk, this seems a good way to manage it... but there is something that dosn't work.

I updated the Locale.php to test locally, and then I call the new method in the boot() function of my plugin in this way for example:
_\RainLab\Translate\Models\Locale::forceDefault('**xx**');_

But in the rewriteTranslatablePageUrl I still get "**en**" for locale and for translatableDefault, so the condition is false and the execution of this command **$localeUrl = $this->getSettingsUrlAttributeTranslated($locale) ?: $localeUrl** is skipped. 
Instead we should force that operation.
Am I forgetting something?

And another point... I just noticed that I should "force" the same behaviour also for the title of the page (so, taking also the localeTitle[]).

 May you check it again please? thank you!

**Comment by daftspunk** [comment]
@ivang76 [Please join slack](https://octobercms.com/forum/post/octobercms-is-now-on-slack) and we can discuss further.

**Comment by daftspunk** [comment]
@ivang76 I have created an issue #636 for your edge case

**Comment by daftspunk** [comment]
This design issue is addressed in v3.1 and is now included as part of the core multisite feature, where the locale is now abstracted into a site instead.

- Link to the [upgrade guide](https://github.com/rainlab/translate-plugin/blob/master/UPGRADE.md)
- Link to [multisite documentation](https://docs.octobercms.com/3.x/cms/resources/multisite.html)

Remember that changing any site's locale will require retranslation since the existing messages are associated with that site. This is expected behavior, and if this becomes a common problem, we will look at introducing a console command to swap messages between sites and locales.

Get in touch with the team for further advice: https://octobercms.com/contact

Thank you
